library/common/thread/lock_guard.h

```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+
+#include "envoy/thread/thread.h"
+
+namespace Envoy {
+namespace Thread {
+
+/**
+ * A lock guard that deals with an optional lock and allows call-sites to release the lock prior to
+ * the lock guard going out of scope.
+ */
+// TODO(junr03): this could be moved to Envoy's codebase.
+class ABSL_SCOPED_LOCKABLE OptionalReleasableLockGuard {
+public:
+  /**
+   * Establishes a scoped mutex-lock. If non-null, the mutex is locked upon construction.
+   *
+   * @param lock the mutex.
+   */
+  OptionalReleasableLockGuard(BasicLockable* lock) ABSL_EXCLUSIVE_LOCK_FUNCTION(lock)
```

Comment:
Cool.

Reply:
Yeah I think LockGuards are super neat one of my favorite RAII-enabled patterns.